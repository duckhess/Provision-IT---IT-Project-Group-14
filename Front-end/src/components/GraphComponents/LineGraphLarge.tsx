import { ResponsiveContainer, LineChart, CartesianGrid, Line, XAxis, YAxis, Tooltip, Legend } from 'recharts';

type Unit = "%" | "$" | "days" | "Benchmark" | "Times" | "Ratio";
import type { Metric } from "../../Types/Types";


interface Dataset {
  name: string; // label
  data: { x: number; y: number }[];
  metric: Metric;
  unit: Unit;
}

interface GraphProps {
  datasets: Dataset[]; // up to 4 datasets
  mergedSets: Dataset[];
  yLabel: String;
  title: String;
}

const colors = ["#8884d8", "#82ca9d", "#ffc658", "#ff7300"];

const LineGraphLarge = ({ datasets, mergedSets, yLabel, title }: GraphProps) => {
  return (

    
<div className="flex flex-col items-start w-[100%] h-[800px] bg-gray-100 rounded-lg shadow p-4">
  <div className ="px-4 w-full">
      <h2 className='text-black text-xl font-bold border-b mb-4 inline-block'>
        {title}
      </h2>
    </div>
  <div className="w-full h-full">
    <ResponsiveContainer width="100%" height="100%">
      <LineChart data={mergedSets}>
        <CartesianGrid stroke="#ccc" />
        <XAxis dataKey="x" />
        <YAxis tickFormatter={(v) => `${v}${yLabel}`} />
        <Tooltip />
        <Legend />
        {datasets.map((ds, i) => (
          <Line
            key={i}
            type="monotone"
            dataKey={ds.name}
            stroke={colors[i % colors.length]}
            strokeWidth={i + 1}
            strokeDasharray="5 5"
          />
        ))}
      </LineChart>
    </ResponsiveContainer>
  </div>
</div>
  )
}

export default LineGraphLarge