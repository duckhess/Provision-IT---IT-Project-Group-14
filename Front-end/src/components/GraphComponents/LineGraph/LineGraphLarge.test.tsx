import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import LineGraphLarge from "./LineGraphLarge";
import type { Dataset } from "../../Types/Types";

// ðŸ”¹ Local array used to collect <Line> props
const renderedLines: any[] = [];

// --- Mock Recharts components --- //
vi.mock("recharts", async (importOriginal) => {
  const original: any = await importOriginal();
  return {
    ...original,
    ResponsiveContainer: ({ children }: any) => <div>{children}</div>,
    LineChart: ({ children }: any) => <div>{children}</div>,
    CartesianGrid: () => <div />,
    XAxis: () => <div />,
    YAxis: () => <div />,
    Tooltip: () => <div />,
    Legend: () => (
      <div data-testid="mock-legend">
        <span>Operating Cycle</span>
        <span>Quick Ratio (Acid Test)</span>
      </div>
    ),
    Line: (props: any) => {
      renderedLines.push(props);
      return <div data-testid={`line-${props.dataKey}`} />;
    },
  };
});

describe("LineGraphLarge", () => {
  const mockDatasets: Dataset[] = [
    {
      name: "Operating Cycle",
      data: [
        { x: 2023, y: 3.28 },
        { x: 2024, y: 3.12 },
      ],
      metric: "Key Ratios",
      unit: "days",
    },
    {
      name: "Quick Ratio (Acid Test)",
      data: [
        { x: 2023, y: 1.87 },
        { x: 2024, y: 1.39 },
      ],
      metric: "Key Ratios",
      unit: "days",
    },
  ];

  const mockMergedSets = [
    { x: 2023, "Operating Cycle": 3.28, "Quick Ratio (Acid Test)": 1.87 },
    { x: 2024, "Operating Cycle": 3.12, "Quick Ratio (Acid Test)": 1.39 },
  ];

  beforeEach(() => {
    renderedLines.length = 0; // reset mock data before each test
  });

  it("renders the chart title", () => {
    render(
      <LineGraphLarge
        datasets={mockDatasets}
        mergedSets={mockMergedSets}
        yLabel="days"
        title="Key Ratios"
      />
    );

    expect(screen.getByText("Key Ratios")).toBeInTheDocument();
  });

  it("renders a line for each dataset with correct dataKeys", () => {
    render(
      <LineGraphLarge
        datasets={mockDatasets}
        mergedSets={mockMergedSets}
        yLabel="days"
        title="Key Ratios"
      />
    );

    const renderedKeys = renderedLines.map((line) => line.dataKey);
    const expectedKeys = mockDatasets.map((ds) => ds.name);

    expect(renderedKeys).toEqual(expectedKeys);

    mockDatasets.forEach((ds) => {
      expect(screen.getByTestId(`line-${ds.name}`)).toBeInTheDocument();
    });
  });

  it("renders legend items", () => {
    render(
      <LineGraphLarge
        datasets={mockDatasets}
        mergedSets={mockMergedSets}
        yLabel="days"
        title="Key Ratios"
      />
    );

    expect(screen.getByText("Operating Cycle")).toBeInTheDocument();
    expect(screen.getByText("Quick Ratio (Acid Test)")).toBeInTheDocument();
  });

  it("matches snapshot", () => {
    const { container } = render(
      <LineGraphLarge
        datasets={mockDatasets}
        mergedSets={mockMergedSets}
        yLabel="days"
        title="Key Ratios"
      />
    );
    expect(container).toMatchSnapshot();
  });
});
