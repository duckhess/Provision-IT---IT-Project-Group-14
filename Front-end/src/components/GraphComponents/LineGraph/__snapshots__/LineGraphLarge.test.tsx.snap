// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LineGraphLarge > matches snapshot 1`] = `
<div>
  <div
    class="flex flex-col items-start w-[100%] h-[800px] bg-gray-100 rounded-lg shadow p-4"
  >
    <div
      class="px-4 w-full"
    >
      <h2
        class="text-black text-xl font-bold border-b mb-4 inline-block"
      >
        Key Ratios
      </h2>
    </div>
    <div
      class="w-full h-full"
    >
      <div>
        <div>
          <div />
          <div />
          <div />
          <div />
          <div
            data-testid="mock-legend"
          >
            <span>
              Operating Cycle
            </span>
            <span>
              Quick Ratio (Acid Test)
            </span>
          </div>
          <div
            data-testid="line-Operating Cycle"
          />
          <div
            data-testid="line-Quick Ratio (Acid Test)"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
